lib/ModuleManager.cpp

```diff
@@ -466,22 +466,80 @@ void QoreModuleManager::init(bool se) {
       QoreModuleManager::addStandardModulePaths();
 }
 
+/* this internal helper function solves an issue with modules
+   on MS Windows (or maybe OSX later).
+
+   The problem with Windows is that we are distributing zip
+   files which can be extracted anywhere. But these binaries
+   had hardcoded paths for modules (/z/something/something or
+   C:/msys64/home/pvanek/src/qore/RELEASE/share/qore-modules
+   for example). These hardcoded paths are fine on Linux because
+   the language should be distributed in a form of system packages
+   (rpm, deb). But on Windows the user decides where to put
+   (extract) the zip archive content. It can be anywhere.
+
+   Of course there is QORE_MODULE_DIR env variable. But it
+   requires logout/login on Windows to be in action. Also
+   user has to set 4 paths there. And unfortunately 2 of
+   them must contain qore version in the path. So it
+   required to change env variable (and re-login) after any
+   qore upgrade. It was annoying.
+
+   So for now - if there is MODULES_RELATIVE_PATH defined
+   from cmake (WIN builds only for now) the location of
+   module(s) is constructed from the location of qore binary.
+ */
+#ifdef MODULES_RELATIVE_PATH
+#ifdef _Q_WINDOWS
+#include <windows.h>
+#endif
+#endif
+const char * module_dir_prefix(const char * path) {
+#ifdef MODULES_RELATIVE_PATH
+#ifdef _Q_WINDOWS
+    // get windows qore.exe binary
+    // Full path may be longer than MAX_PATH
+    // Let's expand it until it fits
+    std::vector<char> buff;
+    DWORD copied = 0;
+    do {
+        buff.resize(buff.size() + MAX_PATH);
+        copied = GetModuleFileName(0, &buff.at(0), buff.size());
+    } while (copied >= buff.size());
+
+    buff.resize(copied);
+    std::string prefix_path(buff.data(), buff.size());
```

Comment:
whát? this looks really really ugly, you are resizing vector of chars and continuously trying to getModuleFileName over and over? where is the `GetModuleFileName` I cannot find it, cannot it be updated to match better?

Reply:
It's a Windows system call. 
You should remember that MAX_PATH is often not enough, especially on modern systems with long pathnames.
https://stackoverflow.com/a/33613252

Reply:
1, you can use string, instead of vector, (no additional copy, no need for resizing at the end etc)
```
    std::string prefix_path;
    DWORD copied = 0;
    do {
        prefix_path.resize(prefix_path.size() + MAX_PATH);
        copied = GetModuleFileName(0, &prefix_path.at(0), prefix_path.size());
    } while (copied >= buff.size());
```

2, this is a win api, that uses `LPSTR`, `DWORD` and we are passing `char *`, `size_t` and then we are comparing `dword` with `size_t`, doesn't it raises some warnings? I don't know if it's okay (maybe some issues with multibyte chars? I don't know how lpstr handles them)

3,
> If the function fails, the return value is 0 (zero). 

maybe we should check for it?

Reply:
1, what do you tihnk?
2, ok, proably we can omit it, should be more or less safe
3, I would add a check that would raise an exception?